pkg/codegen/pcl/component.go

```diff
@@ -15,19 +15,76 @@
 package pcl
 
 import (
+	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/hclsyntax"
 	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model"
 )
 
-// Component represents a component definition in a program.
-//
-// TODO(pdg): implement
+// Component represents a component reference in a program.
 type Component struct {
-	Syntax *hclsyntax.Block
+	node
+
+	syntax *hclsyntax.Block
+
+	// The name visible to API calls related to the component. Used as the Name argument in component
+	// constructors, and through those calls to RegisterResource. Must not be modified during code
+	// generation to ensure that components are not renamed (deleted and recreated).
+	logicalName string
+
+	// the name of block declaration
+	name string
+
+	// The location of the source for the component.
+	source string
+
+	// the full (absolute) path of the component directory
```

Comment:
I'd love to explore a system where components don't require an absolute path. 

If a component needs an absolute path, then it becomes computer specific. Our testing code will need logic to template exported components before comparison, etc.

Reply:
I think a bound component does not need the absolute path and it can be just a component name but I was using this for simplicity. I don't get this part:

> Our testing code will need logic to template exported components before comparison, etc.

What do you mean exactly here by testing code, can you please elaborate?

Reply:
Sorry, I was reading this as a spec of what would need to appear in our `.pcl` files. I see that we append the absolute portion when binding.

https://github.com/pulumi/pulumi/blob/0c27ce7613bdc688ce879d27a5ef7715bccc4b27/pkg/codegen/pcl/binder_component.go#L91-L93

Reply:
Yeah I am just passing the `componentSourceDir` through to the bound component and using that as a uniqueness thing when collecting the child programs into a flat list.